components/script/dom/blob.rs

```diff
@@ -88,22 +90,76 @@ impl DataSlice {
 pub struct Blob {
     reflector_: Reflector,
     #[ignore_heap_size_of = "No clear owner"]
-    data: DataSlice,
+    blob_impl: BlobImpl,
     typeString: String,
     isClosed_: Cell<bool>,
 }
 
+#[derive(Clone, JSTraceable)]
+pub struct BlobImpl {
+    slice: DOMRefCell<Option<DataSlice>>,
+    file_id: Option<Uuid>,
+}
+
+impl BlobImpl {
+    pub fn new_from_slice(slice: DataSlice) -> BlobImpl {
+        BlobImpl {
+            slice: DOMRefCell::new(Some(slice)),
+            file_id: None,
+        }
+    }
+
+    pub fn new_from_file(file_id: Uuid) -> BlobImpl {
+        BlobImpl {
+            slice: DOMRefCell::new(None),
+            file_id: Some(file_id),
+        }
+    }
+
+    pub fn new_from_empty_slice() -> BlobImpl {
+        BlobImpl::new_from_slice(DataSlice::empty())
+    }
+
+    pub fn get_bytes(&self) -> Vec<u8> {
+        match *self.slice.borrow() {
```

Comment:
Currently I don't put any read operations in `blob.rs`, which means that the user needs to check if it is cached and then cache if necessary. Or, if the file is not cached, a empty slice or stuff with similar semantics will be returned.